src/main/kotlin/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.kt

```diff
@@ -0,0 +1,525 @@
+                    ret.exchange,
+                    ret.routingKey,
+                    ret.replyCode,
+                    ret.replyText)
+            }
+            channel
+        } catch (e: IOException) {
+            throw IllegalStateException("Can not create channel", e)
+        }
+    }
+
+    private fun waitForConnectionRecovery(notifier: ShutdownNotifier, waitForRecovery: Boolean = true) {
+        if (isConnectionRecovery(notifier)) {
+            if (waitForRecovery) {
+                waitForRecovery(notifier)
+            } else {
+                LOGGER.warn("Skip waiting for connection recovery")
+            }
+        }
+        check(!connectionIsClosed.get()) { "Connection is already closed" }
+    }
+
+    private fun waitForRecovery(notifier: ShutdownNotifier) {
+        LOGGER.warn("Start waiting for connection recovery")
+        while (isConnectionRecovery(notifier)) {
+            try {
+                Thread.sleep(1)
+            } catch (e: InterruptedException) {
+                LOGGER.error("Wait for connection recovery was interrupted", e)
+                break
+            }
+        }
+        LOGGER.info("Stop waiting for connection recovery")
+    }
+
+    private fun isConnectionRecovery(notifier: ShutdownNotifier): Boolean {
+        return !notifier.isOpen && !connectionIsClosed.get()
+    }
+
+    /**
+     * @param channel pass channel witch used for basicConsume, because delivery tags are scoped per channel,
+     * deliveries must be acknowledged on the same channel they were received on.
+     * @throws IOException
+     */
+    @Throws(IOException::class)
+    private fun basicAck(channel: Channel, deliveryTag: Long) {
+        channel.basicAck(deliveryTag, false)
+    }
+
+    fun lockSendingIfSizeLimitExceeded() {
+        sizeCheckExecutor.submit(::internalLockSendingIfSizeLimitExceeded).get()
+    }
+
+    private fun internalLockSendingIfSizeLimitExceeded() = try {
+        val queueNameToSize = client.queues
+            .associateBy({ it.name }, { it.totalMessages })
+        knownExchangesToRoutingKeys.entries.asSequence()
+            .flatMap { (exchange, routingKeys) ->
+                client.getBindingsBySource(rabbitMQConfiguration.vHost, exchange).asSequence()
+                    .filter { it.destinationType == DestinationType.QUEUE && routingKeys.contains(it.routingKey) }
+            }
+            .groupBy { it.routingKey }
+            .forEach { (routingKey, bindings) ->
+                val bindingNameToSize = bindings
+                    .associateBy({ it.destination }, { queueNameToSize.getValue(it.destination) })
+                val limit = connectionManagerConfiguration.virtualPublishLimit
+                val holder = getChannelFor(PinId.forRoutingKey(routingKey))
+                LOGGER.trace { "Size limit lock for routing key '$routingKey': ${holder.sizeLimitLock}" }
```

Comment:
I think it may be helpful, we'll have something like:
```
Size limit lock for routing key 'parsed-key': java.util.concurrent.locks.ReentrantLock@5744e6e4[Unlocked]
Size limit lock for routing key 'parsed-key': java.util.concurrent.locks.ReentrantLock@5744e6e4[Locked by thread pool-3-thread-1]
```